File: frontend/src/calculations/comparison.ts
import type { ComparisonRow, ComparisonStatus, ElementType, Side, StateKind, WidthSample } from "../types/domain";

/** Direct port of backend/app/calculations/comparison.py, extended to a
 * 3-way comparison once a "Projet V1" file is uploaded: Existant → Projet
 * V0 is the original comparison, and Projet V0 → Projet V1 / Existant →
 * Projet V1 are the two additional deltas that come with a second project
 * version. All three are always computed; a pair with a missing state (no
 * V1 uploaded) simply comes back null, same as an empty file always has. */

function interpolate(pk: number, pks: number[], widths: number[]): number | null {
  if (pks.length === 0) return null;
  // bisect_left
  let lo = 0;
  let hi = pks.length;
  while (lo < hi) {
    const mid = (lo + hi) >> 1;
    if (pks[mid] < pk) lo = mid + 1;
    else hi = mid;
  }
  const idx = lo;
  if (idx === 0) return Math.abs(pks[0] - pk) < 1e-6 ? widths[0] : null;
  if (idx >= pks.length) return Math.abs(pk - pks[pks.length - 1]) < 1e-6 ? widths[widths.length - 1] : null;
  if (pks[idx] === pk) return widths[idx];
  const pk0 = pks[idx - 1];
  const pk1 = pks[idx];
  const w0 = widths[idx - 1];
  const w1 = widths[idx];
  if (pk1 === pk0) return w0;
  const t = (pk - pk0) / (pk1 - pk0);
  return w0 + t * (w1 - w0);
}

export function compareStates(samples: WidthSample[], deltaSeuilM: number): ComparisonRow[] {
  type Group = { pks: number[]; widths: number[]; byPk: Map<number, WidthSample> };
  const byGroup = new Map<string, Group>();
  const keyOf = (side: Side, et: ElementType, state: StateKind) => `${side}|${et}|${state}`;

  const sorted = [...samples]
    .filter((s) => s.width_m != null && s.element_type !== "non_utilise")
    .sort((a, b) => a.pk - b.pk);

  for (const s of sorted) {
    const key = keyOf(s.side, s.element_type, s.state);
    if (!byGroup.has(key)) byGroup.set(key, { pks: [], widths: [], byPk: new Map() });
    const g = byGroup.get(key)!;
    g.pks.push(s.pk);
    g.widths.push(s.width_m as number);
    g.byPk.set(s.pk, s);
  }

  const pairKeys = new Set<string>();
  for (const s of sorted) pairKeys.add(`${s.side}|${s.element_type}`);

  const statusOf = (delta: number | null): ComparisonStatus | null => {
    if (delta == null) return null;
    if (delta > deltaSeuilM) return "ameliore";
    if (delta < -deltaSeuilM) return "degrade";
    return "inchange";
  };

  const rows: ComparisonRow[] = [];
  for (const pairKey of Array.from(pairKeys).sort()) {
    const [side, elementType] = pairKey.split("|") as [Side, ElementType];
    const empty: Group = { pks: [], widths: [], byPk: new Map() };
    const existant = byGroup.get(keyOf(side, elementType, "existant")) ?? empty;
    const projetV0 = byGroup.get(keyOf(side, elementType, "projet")) ?? empty;
    const projetV1 = byGroup.get(keyOf(side, elementType, "projet_v1")) ?? empty;
    if (existant.pks.length === 0 && projetV0.pks.length === 0 && projetV1.pks.length === 0) continue;

    const allPks = Array.from(new Set([...existant.pks, ...projetV0.pks, ...projetV1.pks])).sort((a, b) => a - b);
    for (const pk of allPks) {
      const wExist = interpolate(pk, existant.pks, existant.widths);
      const wV0 = interpolate(pk, projetV0.pks, projetV0.widths);
      const wV1 = interpolate(pk, projetV1.pks, projetV1.widths);

      const deltaExistantV0 = wExist != null && wV0 != null ? wV0 - wExist : null;
      const deltaV0V1 = wV0 != null && wV1 != null ? wV1 - wV0 : null;
      const deltaExistantV1 = wExist != null && wV1 != null ? wV1 - wExist : null;

      // Every pk here came verbatim from one state's own sample list, so
      // that sample's own near/far is a real plan position — not a new
      // interpolation — good enough to draw this row "en situation".
      const origin = existant.byPk.get(pk) ?? projetV0.byPk.get(pk) ?? projetV1.byPk.get(pk);
      rows.push({
        pk,
        side,
        element_type: elementType,
        width_existant: wExist,
        width_projet_v0: wV0,
        width_projet_v1: wV1,
        delta_existant_v0: deltaExistantV0,
        status_existant_v0: statusOf(deltaExistantV0),
        delta_v0_v1: deltaV0V1,
        status_v0_v1: statusOf(deltaV0V1),
        delta_existant_v1: deltaExistantV1,
        status_existant_v1: statusOf(deltaExistantV1),
        near_x: origin?.near_x,
        near_y: origin?.near_y,
        far_x: origin?.far_x,
        far_y: origin?.far_y,
      });
    }
  }
  return rows;
}

File: backend/app/calculations/comparison.py
"""Existant vs projet comparison at matching PK, interpolating the projet
width when stations don't coincide exactly."""

from __future__ import annotations

import bisect
from collections import defaultdict

from app.models.domain import ComparisonRow, ComparisonStatus, ElementType, Side, StateKind, WidthSample


def _interpolate(pk: float, pks: list[float], widths: list[float]) -> float | None:
    if not pks:
        return None
    idx = bisect.bisect_left(pks, pk)
    if idx == 0:
        return widths[0] if pks[0] - pk < 1e-6 else None
    if idx >= len(pks):
        return widths[-1] if pk - pks[-1] < 1e-6 else None
    if pks[idx] == pk:
        return widths[idx]
    pk0, pk1 = pks[idx - 1], pks[idx]
    w0, w1 = widths[idx - 1], widths[idx]
    if pk1 == pk0:
        return w0
    t = (pk - pk0) / (pk1 - pk0)
    return w0 + t * (w1 - w0)


def compare_states(samples: list[WidthSample], delta_seuil_m: float) -> list[ComparisonRow]:
    by_group: dict[tuple[Side, ElementType, StateKind], list[WidthSample]] = defaultdict(list)
    for s in samples:
        if s.width_m is not None:
            by_group[(s.side, s.element_type, s.state)].append(s)

    keys = {(side, et) for (side, et, _state) in by_group}
    rows: list[ComparisonRow] = []
    for side, element_type in sorted(keys, key=lambda k: (k[0].value, k[1].value)):
        existant = sorted(by_group.get((side, element_type, StateKind.EXISTANT), []), key=lambda s: s.pk)
        projet = sorted(by_group.get((side, element_type, StateKind.PROJET), []), key=lambda s: s.pk)
        if not existant and not projet:
            continue
        projet_pks = [s.pk for s in projet]
        projet_widths = [s.width_m for s in projet]
        existant_pks = [s.pk for s in existant]
        existant_widths = [s.width_m for s in existant]

        all_pks = sorted({s.pk for s in existant} | {s.pk for s in projet})
        for pk in all_pks:
            w_exist = _interpolate(pk, existant_pks, existant_widths)
            w_proj = _interpolate(pk, projet_pks, projet_widths)
            delta = None
            status = None
            if w_exist is not None and w_proj is not None:
                delta = w_proj - w_exist
                if delta > delta_seuil_m:
                    status = ComparisonStatus.AMELIORE
                elif delta < -delta_seuil_m:
                    status = ComparisonStatus.DEGRADE
                else:
                    status = ComparisonStatus.INCHANGE
            rows.append(
                ComparisonRow(
                    pk=pk,
                    side=side,
                    element_type=element_type,
                    width_existant=w_exist,
                    width_projet=w_proj,
                    delta=delta,
                    status=status,
                )
            )
    return rows
